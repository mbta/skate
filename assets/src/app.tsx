// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
