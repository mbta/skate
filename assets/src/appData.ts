const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
