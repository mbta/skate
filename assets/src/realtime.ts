import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
