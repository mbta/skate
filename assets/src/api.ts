import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
