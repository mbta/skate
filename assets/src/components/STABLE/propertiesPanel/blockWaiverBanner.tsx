import React from "react"
import {
  CurrentFuturePastType,
  currentFuturePastType,
} from "../../../models/blockWaiver"
import { BlockWaiver } from "../../../realtime"
import { formattedTime } from "../../../util/dateTime"
import IconAlertCircle, { AlertIconStyle } from "../../iconAlertCircle"

interface Props {
  blockWaiver: BlockWaiver
}

const currentFuturePastClass = (blockWaiver: BlockWaiver): string => {
  switch (currentFuturePastType(blockWaiver)) {
    case CurrentFuturePastType.Current:
      return "current"
    case CurrentFuturePastType.Future:
      return "future"
    case CurrentFuturePastType.Past:
      return "past"
  }
}

const currentFuturePastTitle = (blockWaiver: BlockWaiver): string => {
  switch (currentFuturePastType(blockWaiver)) {
    case CurrentFuturePastType.Current:
      return "Current"
    case CurrentFuturePastType.Future:
      return "Future"
    case CurrentFuturePastType.Past:
      return "Past"
  }
}

const alertIconStyle = (blockWaiver: BlockWaiver): AlertIconStyle => {
  switch (currentFuturePastType(blockWaiver)) {
    case CurrentFuturePastType.Current:
      return AlertIconStyle.Black
    case CurrentFuturePastType.Future:
      return AlertIconStyle.GreyOnGrey
    case CurrentFuturePastType.Past:
      return AlertIconStyle.GreyOnGrey
  }
}

const BlockWaiverBanner = ({ blockWaiver }: Props) => (
  <div
    className={`m-block-waiver-banner m-block-waiver-banner--${currentFuturePastClass(
      blockWaiver
    )}`}
  >
    <div className="m-block-waiver-banner__header">
      <span className="m-block-waiver-banner__alert-icon">
        <IconAlertCircle style={alertIconStyle(blockWaiver)} />
      </span>
      <div className="m-block-waiver-banner__title">
        Dispatcher Note - {currentFuturePastTitle(blockWaiver)}
      </div>
    </div>

    <table className="m-block-waiver-banner__details">
      <tbody>
        <tr>
          <td className="m-block-waiver-banner__detail-label">Reason</td>
          <td className="m-block-waiver-banner__detail-value">
            {[
              blockWaiver.causeDescription,
              (blockWaiver.remark || "")
                .replace(blockWaiver.causeDescription, "")
                .replace(/:$/, ""),
            ]
              .filter((s) => s !== null && s !== "")
              .join(" ")}
          </td>
        </tr>
        <tr>
          <td className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--start-time">
            Start Time
          </td>
          <td className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--start-time">
            {formattedTime(blockWaiver.startTime)}
          </td>
        </tr>
        <tr>
          <td className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--end-time">
            End Time
          </td>
          <td className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--end-time">
            {formattedTime(blockWaiver.endTime)}
          </td>
        </tr>
      </tbody>
    </table>
  </div>
)

export default BlockWaiverBanner
