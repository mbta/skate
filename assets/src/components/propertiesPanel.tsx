import React, { useContext, useEffect } from "react"
import { StateDispatchContext } from "../contexts/stateDispatchContext"
import detectSwipe, { SwipeDirection } from "../helpers/detectSwipe"
import { isVehicle } from "../models/vehicle"
import { VehicleOrGhost } from "../realtime.d"
import { Route } from "../schedule"
import { deselectVehicle } from "../state"
import CloseButton from "./propertiesPanel/closeButton"
import GhostPropertiesPanel from "./propertiesPanel/ghostPropertiesPanel"
import VehiclePropertiesPanel from "./propertiesPanel/vehiclePropertiesPanel"

interface Props {
  selectedVehicleOrGhost: VehicleOrGhost
  route?: Route
}

export const handleSwipe = (hideMe: () => void) => (
  swipeDirection: SwipeDirection,
  target: HTMLElement | null
) => {
  if (target && target.id === "id-vehicle-map") {
    return
  }

  if (swipeDirection === "Right") {
    hideMe()
  }
}

const PropertiesPanel = ({ selectedVehicleOrGhost, route }: Props) => {
  const [, dispatch] = useContext(StateDispatchContext)

  const hideMe = () => dispatch(deselectVehicle())

  useEffect(() => {
    return detectSwipe("m-properties-panel", handleSwipe(hideMe))
  })

  return (
    <>
      <div id="m-properties-panel" className="m-properties-panel">
        <div className="m-properties-panel__vehicle-or-ghost-panel">
          {isVehicle(selectedVehicleOrGhost) ? (
            <VehiclePropertiesPanel
              selectedVehicle={selectedVehicleOrGhost}
              route={route}
            />
          ) : (
            <GhostPropertiesPanel
              selectedGhost={selectedVehicleOrGhost}
              route={route}
            />
          )}
        </div>

        <CloseButton />
      </div>
      <div className="m-properties-panel__modal-overlay" onClick={hideMe} />
    </>
  )
}

export default PropertiesPanel
