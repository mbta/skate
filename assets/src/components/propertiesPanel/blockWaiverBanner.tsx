import React from "react"
import ReactTooltip from "react-tooltip"
import { questionMarkIcon } from "../../helpers/icon"
import {
  CurrentFuturePastType,
  currentFuturePastType,
} from "../../models/blockWaiver"
import { BlockWaiver } from "../../realtime"
import { formattedTime } from "../../util/dateTime"
import IconAlertCircle, { AlertIconStyle } from "../iconAlertCircle"

interface Props {
  blockWaiver: BlockWaiver
}

const currentFuturePastClass = (blockWaiver: BlockWaiver): string => {
  switch (currentFuturePastType(blockWaiver)) {
    case CurrentFuturePastType.Current:
      return "current"
    case CurrentFuturePastType.Future:
      return "future"
    case CurrentFuturePastType.Past:
      return "past"
  }
}

const currentFuturePastTitle = (blockWaiver: BlockWaiver): string => {
  switch (currentFuturePastType(blockWaiver)) {
    case CurrentFuturePastType.Current:
      return "Current"
    case CurrentFuturePastType.Future:
      return "Future"
    case CurrentFuturePastType.Past:
      return "Past"
  }
}

const alertIconStyle = (blockWaiver: BlockWaiver): AlertIconStyle => {
  switch (currentFuturePastType(blockWaiver)) {
    case CurrentFuturePastType.Current:
      return AlertIconStyle.Black
    case CurrentFuturePastType.Future:
      return AlertIconStyle.GreyOnGrey
    case CurrentFuturePastType.Past:
      return AlertIconStyle.GreyOnGrey
  }
}

const ReasonTooltip = () => (
  <ReactTooltip
    id="reason"
    className="m-block-waiver-banner__reason-tooltip"
    place="bottom"
    effect="solid"
    globalEventOff="click"
  >
    <h3>Reason codes</h3>

    <ul>
      <li>
        <strong>B</strong> - Manpower
      </li>
      <li>
        <strong>C</strong> - No equipment
      </li>
      <li>
        <strong>D</strong> - Disabled bus
      </li>
      <li>
        <strong>E</strong> - Diverted to other work
      </li>
      <li>
        <strong>F</strong> - Traffic
      </li>
      <li>
        <strong>G</strong> - Accident
      </li>
      <li>
        <strong>H</strong> - Weather
      </li>
      <li>
        <strong>I</strong> - Operator error
      </li>
      <li>
        <strong>J</strong> - Other
      </li>
    </ul>
  </ReactTooltip>
)

const BlockWaiverBanner = ({ blockWaiver }: Props) => (
  <div
    className={`m-block-waiver-banner m-block-waiver-banner--${currentFuturePastClass(
      blockWaiver
    )}`}
  >
    <div className="m-block-waiver-banner__header">
      <span className="m-block-waiver-banner__alert-icon">
        <IconAlertCircle style={alertIconStyle(blockWaiver)} />
      </span>
      <div className="m-block-waiver-banner__title">
        Dispatcher Note - {currentFuturePastTitle(blockWaiver)}
      </div>
    </div>

    <table className="m-block-waiver-banner__details">
      <tbody>
        <tr>
          <td
            className="m-block-waiver-banner__detail-label m-block-waiver-banner__reason"
            data-tip={true}
            data-for="reason"
            data-event="click"
          >
            Reason {questionMarkIcon("m-block-waiver-banner__reason-icon")}
          </td>
          <td className="m-block-waiver-banner__detail-value">
            {blockWaiver.remark}
          </td>
        </tr>
        <tr>
          <td className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--start-time">
            Start Time
          </td>
          <td className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--start-time">
            {formattedTime(blockWaiver.startTime)}
          </td>
        </tr>
        <tr>
          <td className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--end-time">
            End Time
          </td>
          <td className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--end-time">
            {formattedTime(blockWaiver.endTime)}
          </td>
        </tr>
      </tbody>
    </table>

    <ReasonTooltip />
  </div>
)

export default BlockWaiverBanner
