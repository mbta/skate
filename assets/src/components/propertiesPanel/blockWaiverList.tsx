import React from "react"
import { BlockWaiver } from "../../realtime"
import BlockWaiverBanner from "./blockWaiverBanner"

interface Props {
  blockWaivers: BlockWaiver[]
}

const BlockWaiverList = ({ blockWaivers }: Props) => (
  <div className="m-block-waiver-list">
    {blockWaivers.map(blockWaiver => (
      <BlockWaiverBanner
        blockWaiver={blockWaiver}
        key={`${blockWaiver.startTime}-${blockWaiver.endTime}`}
      />
    ))}
  </div>
)

export default BlockWaiverList
