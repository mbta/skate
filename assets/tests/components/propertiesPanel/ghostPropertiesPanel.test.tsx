import React from "react"
import renderer from "react-test-renderer"
import GhostPropertiesPanel from "../../../src/components/propertiesPanel/ghostPropertiesPanel"
import { BlockWaiver, Ghost } from "../../../src/realtime"
import { Route } from "../../../src/schedule"
import * as blockWaiverBanner from "../../../src/components/propertiesPanel/blockWaiverBanner"

jest.spyOn(blockWaiverBanner, "nowEpochSeconds").mockImplementation(() => 81720)

// Enable feature flags
jest.mock("../../../src/laboratoryFeatures", () => ({
  __esModule: true,
  default: jest.fn(() => true),
}))

const ghost: Ghost = {
  id: "ghost-trip",
  directionId: 0,
  routeId: "39",
  tripId: "trip",
  headsign: "headsign",
  blockId: "block",
  runId: "123-0123",
  viaVariant: "X",
  layoverDepartureTime: null,
  scheduledTimepointStatus: {
    timepointId: "t0",
    fractionUntilTimepoint: 0.0,
  },
  routeStatus: "on_route",
  blockWaivers: [],
}

const route: Route = {
  id: "39",
  directionNames: {
    0: "Outbound",
    1: "Inbound",
  },
  name: "39",
}

describe("GhostPropertiesPanel", () => {
  test("renders", () => {
    const tree = renderer
      .create(<GhostPropertiesPanel selectedGhost={ghost} route={route} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders ghost with missing run", () => {
    const ghostWithoutRun: Ghost = { ...ghost, runId: null }
    const tree = renderer
      .create(
        <GhostPropertiesPanel selectedGhost={ghostWithoutRun} route={route} />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders ghost with block waivers", () => {
    const blockWaiver: BlockWaiver = {
      startTime: 18300,
      endTime: 45480,
      remark: "E:1106",
    }
    const ghostWithBlockWaivers: Ghost = {
      ...ghost,
      blockWaivers: [blockWaiver],
    }

    const tree = renderer
      .create(<GhostPropertiesPanel selectedGhost={ghostWithBlockWaivers} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
