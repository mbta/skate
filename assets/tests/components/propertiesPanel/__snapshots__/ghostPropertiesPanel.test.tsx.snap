// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GhostPropertiesPanel renders 1`] = `
<div
  className="m-ghost-properties-panel"
>
  <div
    className="m-properties-panel__header"
  >
    <div
      className="m-properties-panel__label"
    >
      <svg
        style={
          Object {
            "height": 64,
            "width": 64,
          }
        }
        viewBox="-32 -20 64 64"
      >
        <g
          className="m-vehicle-icon m-vehicle-icon--large ghost"
        >
          <rect
            className="m-vehicle-icon__label-background"
            height={24}
            rx={12}
            ry={12}
            width={64}
            x={-32}
            y={19}
          />
          <text
            className="m-vehicle-icon__label m-vehicle-icon__label--normal"
            dominantBaseline="central"
            textAnchor="middle"
            x="0"
            y={31}
          >
            0123
          </text>
          <g
            transform="scale(0.7) translate(-24,-23)"
          >
            <path
              className="m-vehicle-icon__ghost-highlight"
              d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
              stroke-join="round"
            />
            <path
              className="m-vehicle-icon__ghost-body"
              d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
              stroke-join="round"
            />
          </g>
          <text
            className="m-vehicle-icon__variant"
            dominantBaseline="alphabetic"
            textAnchor="middle"
            x={0}
            y={8}
          >
            X
          </text>
        </g>
      </svg>
    </div>
    <div
      className="m-properties-panel__variant"
    >
      <div
        className="m-properties-panel__inbound-outbound"
      >
        Outbound
      </div>
      <div
        className="m-route-variant-name"
      >
        <span
          className="m-route-variant-name__route-id"
        >
          39_X
        </span>
         headsign
      </div>
    </div>
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
  </div>
  <div
    className="m-properties-list"
  >
    <table
      className="m-properties-list__table"
    >
      <tbody>
        <tr
          className="m-properties-list__property "
        >
          <td
            className="m-properties-list__property-label"
          >
            Run
          </td>
          <td
            className="m-properties-list__property-value"
          >
            123-0123
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`GhostPropertiesPanel renders ghost with block waivers 1`] = `
<div
  className="m-ghost-properties-panel"
>
  <div
    className="m-properties-panel__header"
  >
    <div
      className="m-properties-panel__label"
    >
      <svg
        style={
          Object {
            "height": 64,
            "width": 64,
          }
        }
        viewBox="-32 -20 64 64"
      >
        <g
          className="m-vehicle-icon m-vehicle-icon--large ghost"
        >
          <rect
            className="m-vehicle-icon__label-background"
            height={24}
            rx={12}
            ry={12}
            width={64}
            x={-32}
            y={19}
          />
          <text
            className="m-vehicle-icon__label m-vehicle-icon__label--normal"
            dominantBaseline="central"
            textAnchor="middle"
            x="0"
            y={31}
          >
            0123
          </text>
          <g
            transform="scale(0.7) translate(-24,-23)"
          >
            <path
              className="m-vehicle-icon__ghost-highlight"
              d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
              stroke-join="round"
            />
            <path
              className="m-vehicle-icon__ghost-body"
              d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
              stroke-join="round"
            />
          </g>
          <text
            className="m-vehicle-icon__variant"
            dominantBaseline="alphabetic"
            textAnchor="middle"
            x={0}
            y={8}
          >
            X
          </text>
        </g>
      </svg>
    </div>
    <div
      className="m-properties-panel__variant"
    >
      <div
        className="m-properties-panel__inbound-outbound"
      >
        
      </div>
      <div
        className="m-route-variant-name"
      >
        <span
          className="m-route-variant-name__route-id"
        >
          39_X
        </span>
         headsign
      </div>
    </div>
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
  </div>
  <div
    className="m-block-waiver-list"
  >
    <div
      className="m-block-waiver-banner m-block-waiver-banner--past"
    >
      <div
        className="m-block-waiver-banner__header"
      >
        <span
          className="m-block-waiver-banner__alert-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        <div
          className="m-block-waiver-banner__title"
        >
          Dispatcher Note - 
          Past
        </div>
      </div>
      <table
        className="m-block-waiver-banner__details"
      >
        <tbody>
          <tr>
            <td
              className="m-block-waiver-banner__detail-label"
            >
              Reason
            </td>
            <td
              className="m-block-waiver-banner__detail-value"
            >
              E:1106
            </td>
          </tr>
          <tr>
            <td
              className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--start-time"
            >
              Start Time
            </td>
            <td
              className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--start-time"
            >
              5:05am
            </td>
          </tr>
          <tr>
            <td
              className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--end-time"
            >
              End Time
            </td>
            <td
              className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--end-time"
            >
              12:38pm
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="m-properties-list"
  >
    <table
      className="m-properties-list__table"
    >
      <tbody>
        <tr
          className="m-properties-list__property "
        >
          <td
            className="m-properties-list__property-label"
          >
            Run
          </td>
          <td
            className="m-properties-list__property-value"
          >
            123-0123
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`GhostPropertiesPanel renders ghost with missing run 1`] = `
<div
  className="m-ghost-properties-panel"
>
  <div
    className="m-properties-panel__header"
  >
    <div
      className="m-properties-panel__label"
    >
      <svg
        style={
          Object {
            "height": 64,
            "width": 64,
          }
        }
        viewBox="-32 -20 64 64"
      >
        <g
          className="m-vehicle-icon m-vehicle-icon--large ghost"
        >
          <rect
            className="m-vehicle-icon__label-background"
            height={24}
            rx={12}
            ry={12}
            width={64}
            x={-32}
            y={19}
          />
          <text
            className="m-vehicle-icon__label m-vehicle-icon__label--normal"
            dominantBaseline="central"
            textAnchor="middle"
            x="0"
            y={31}
          >
            N/A
          </text>
          <g
            transform="scale(0.7) translate(-24,-23)"
          >
            <path
              className="m-vehicle-icon__ghost-highlight"
              d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
              stroke-join="round"
            />
            <path
              className="m-vehicle-icon__ghost-body"
              d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
              stroke-join="round"
            />
          </g>
          <text
            className="m-vehicle-icon__variant"
            dominantBaseline="alphabetic"
            textAnchor="middle"
            x={0}
            y={8}
          >
            X
          </text>
        </g>
      </svg>
    </div>
    <div
      className="m-properties-panel__variant"
    >
      <div
        className="m-properties-panel__inbound-outbound"
      >
        Outbound
      </div>
      <div
        className="m-route-variant-name"
      >
        <span
          className="m-route-variant-name__route-id"
        >
          39_X
        </span>
         headsign
      </div>
    </div>
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
  </div>
  <div
    className="m-properties-list"
  >
    <table
      className="m-properties-list__table"
    >
      <tbody>
        <tr
          className="m-properties-list__property "
        >
          <td
            className="m-properties-list__property-label"
          >
            Run
          </td>
          <td
            className="m-properties-list__property-value"
          >
            Not Available
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
